app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
+    Object.keys(value).forEach((valueKey) => {
+      weakReffedValueObj[valueKey] = new WeakRef(value[valueKey]);
+    });
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
+    const weakReffedValue = this.map.get(key);
+    if (!weakReffedValue) {
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
+    for (const weakReffedValueKey of Object.keys(weakReffedValue)) {
+      const deref = weakReffedValue[weakReffedValueKey].deref();
+      if (deref === undefined) {
+        this.map.delete(weakReffedValueKey);
+        return undefined;
+      }
+      deReffedValue[weakReffedValueKey] = deref;
+    }
+
+    return deReffedValue;
+  }
+
+  has(key: string) {
+    return this.get(key) !== undefined;
+  }
+
+  delete(key: string) {
+    const value = this.get(key);
+    if (value !== undefined) {
+      return this.map.delete(key);
+    }
+    return false;
+  }
+
+  clear() {
+    this.map.clear();
+  }
+
+  forEach(
+    callback: (
+      value: WeakRefCompatibleObject<Record<string, object>>,
+      key: string,
+      map: Map<string, WeakRefCompatibleObject<Record<string, object>>>,
+    ) => void,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```

Comment:
Nit: Typically we encourage adding a comment each time we use `any` for some unavoidable reason, as we are trying to discourage its use. A comment can help explain why it's necessary so someone else doesn't waste time trying to remove it without understanding why.

In this case, I think `any` makes sense because this is an unbound method, so the `this` value is unknown, it can be anything. Also the `Map` type this is based on uses `any` for this parameter as well.

Reply:
yep makes sense! Added a comment here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/a3770013832099fc91e51e849f3ff399b556d4f8#diff-4d13f1f0ba4d949f6dfe026c7c50a982dfd0e3b43d3bbd265f6acdca8f501302R180